### PR TITLE
[hip] Remove deprecated hip APIs, simplify urContext

### PR DESCRIPTION
**Pull request by JackAKirk**
hip amd does not have a native context type other than to support the cuContext type on Nvidia devices. hip plans to completely remove usage of this type anyway, with all corresponding APIs marked deprecated.

This PR future-proofs our hip implementation for when these deprecated APIs become unsupported. It also simplifies the urContext type.

This updates a context test to pass if the interface is unsupported, which cleans up the native_cpu match file that doesn't support certain context interfaces.

### CODE REVIEW

**Comment by JackAKirk** [comment]
intel/llvm tested with these changes here: https://github.com/intel/llvm/pull/14476

File: source/adapters/hip/device.hpp
Comment (frasercrmck) [inline source/adapters/hip/device.hpp:98]: Looks like this was copied from an older comment but we have `HIP` and `hip` in the same comment. I think `HIP` is best?

**Review by EwanC — APPROVED** [review]
command-buffer change LGTM

File: source/adapters/hip/device.hpp
Comment (JackAKirk) [reply source/adapters/hip/device.hpp]: Thanks, I've used HIP consistently now.

**Review by kbenzie — APPROVED** [review]

File: source/adapters/hip/context.hpp
Comment (hdelan) [inline source/adapters/hip/context.hpp:36]: Why are you removing this comment? I think it's quite informative

File: source/adapters/hip/context.hpp
Comment (hdelan) [inline source/adapters/hip/context.hpp:88]: Why no longer retaining ur devices?

File: source/adapters/hip/device.hpp
Comment (hdelan) [inline source/adapters/hip/device.hpp:107]: Nice

**Review by hdelan — APPROVED** [review]
LGTM with a few minor questions.

File: source/adapters/hip/context.hpp
Comment (JackAKirk) [reply source/adapters/hip/context.hpp]: I've added it back.

**Comment by omarahmed1111** [comment]
@JackAKirk Please make sure https://github.com/intel/llvm/pull/14476 is updated and passing CI, so we could merge this. Thanks!

**Comment by kbenzie** [comment]
I've removed *ready to merge* since the intel/llvm PR isn't passing testing and is still awaiting approvals. Once these are in place, add the *ready to merge* label again.

**Comment by JackAKirk** [comment]
This updates a context test to pass if the interface is unsupported, which cleans up the native_cpu match file that doesn't support certain context interfaces.

**Comment by JackAKirk** [comment]
> I've removed _ready to merge_ since the intel/llvm PR isn't passing testing and is still awaiting approvals. Once these are in place, add the _ready to merge_ label again.

Thanks, tests for https://github.com/intel/llvm/pull/14476 pass now and has approval, so I've put back ready to merge. I'll update that PR with the commit hash once this PR is merged.

**Pull request merged**
